pkg/kubelet/dockertools/manager.go

```diff
@@ -1791,6 +1865,43 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubec
 		}
 	}
 
+	// only the last execution of an init container should be preserved, and only preserve it if it is in the
+	// list of init containers to keep.
+	initContainerNames := sets.NewString()
+	for _, container := range pod.Spec.InitContainers {
+		initContainerNames.Insert(container.Name)
+	}
+	for name := range initContainerNames {
+		count := 0
+		for _, status := range podStatus.ContainerStatuses {
+			if !initContainerNames.Has(status.Name) || status.State != kubecontainer.ContainerStateExited {
+				continue
+			}
+			count++
+			if count == 1 {
+				continue
+			}
+			if _, ok := containerChanges.InitContainersToKeep[kubecontainer.DockerID(status.ID.ID)]; ok {
+				continue
+			}
+			glog.V(2).Infof("Removing init container %q instance %q %d", status.Name, status.ID.ID, count)
+			if err := dm.client.RemoveContainer(docker.RemoveContainerOptions{ID: status.ID.ID, RemoveVolumes: true}); err != nil {
+				if _, ok := err.(*docker.NoSuchContainer); ok {
```

Comment:
nit: I'd write a function that just returns the one true status and a list of toDelete, and remove all the toDeletes, instead of the counter